js/modules/k6/grpc/client.go

```diff
@@ -0,0 +1,402 @@
+	if _, ok := c.tags["name"]; !ok && state.Options.SystemTags.Has(stats.TagName) {
+		c.tags["name"] = method
+	}
+
+	reqdm := dynamic.NewMessage(md.GetInputType())
+	b, _ := req.ToObject(rt).MarshalJSON()
+	_ = reqdm.UnmarshalJSON(b)
+
+	reqCtx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	s := grpcdynamic.NewStub(c.conn)
+	header, trailer := metadata.New(nil), metadata.New(nil)
+	resp, err := s.InvokeRpc(reqCtx, md, reqdm, grpc.Header(&header), grpc.Trailer(&trailer))
+
+	var response Response
+	response.Headers = header
+	response.Trailers = trailer
+
+	if err != nil {
+		st := status.Convert(err)
+		response.Status = st.Code()
+		errdm, _ := dynamic.AsDynamicMessage(st.Proto())
+
+		// (rogchap) there is a lot of marshaling/unmarshaling here, but because this is a dynamic message
+		// we need to marshal to get the JSON representation first. Using a map seems the best way to create
+		// a goja.Value from the raw JSON bytes.
+		raw, _ := errdm.MarshalJSON()
+		errMsg := make(map[string]interface{})
+		_ = json.Unmarshal(raw, &errMsg)
+		response.Error = rt.ToValue(errMsg)
+	}
+
+	if resp != nil {
+		msgdm := dynamic.NewMessage(md.GetOutputType())
+		msgdm.Merge(resp)
+		raw, _ := msgdm.MarshalJSON()
+		msg := make(map[string]interface{})
+		_ = json.Unmarshal(raw, &msg)
+		response.Message = rt.ToValue(msg)
+	}
+
+	return &response, nil
+}
+
+// Close will close the client gRPC connection
+func (c *Client) Close() error {
+	if c == nil || c.conn == nil {
+		return nil
+	}
+	err := c.conn.Close()
+	c.conn = nil
+
+	return err
+}
+
+// TagConn implements the stats.Handler interface
+func (*Client) TagConn(ctx context.Context, _ *grpcstats.ConnTagInfo) context.Context {
+	// noop
+	return ctx
+}
+
+// HandleConn implements the stats.Handler interface
+func (*Client) HandleConn(context.Context, grpcstats.ConnStats) {
+	// noop
+}
+
+// TagRPC implements the stats.Handler interface
+func (*Client) TagRPC(ctx context.Context, _ *grpcstats.RPCTagInfo) context.Context {
+	// noop
+	return ctx
+}
+
+// HandleRPC implements the stats.Handler interface
+func (c *Client) HandleRPC(ctx context.Context, stat grpcstats.RPCStats) {
+	state := lib.GetState(ctx)
+
+	switch s := stat.(type) {
+	case *grpcstats.OutHeader:
+		if state.Options.SystemTags.Has(stats.TagIP) && s.RemoteAddr != nil {
+			if ip, _, err := net.SplitHostPort(s.RemoteAddr.String()); err == nil {
+				c.tags["ip"] = ip
+			}
+		}
+	case *grpcstats.End:
+
+		if state.Options.SystemTags.Has(stats.TagStatus) {
+			c.tags["status"] = strconv.Itoa(int(status.Code(s.Error)))
+		}
+
+		tags := stats.IntoSampleTags(&c.tags)
+		stats.PushIfNotDone(ctx, state.Samples, stats.ConnectedSamples{
+			Samples: []stats.Sample{
+				{
+					Metric: metrics.GRPCReqDuration,
+					Tags:   tags,
+					Value:  stats.D(s.EndTime.Sub(s.BeginTime)),
+				},
+				{
+					Metric: metrics.GRPCReqs,
+					Tags:   tags,
+					Value:  1,
+				},
+			},
+		})
```

Comment:
Just a thing that I found out while doing a very fast test:
metrics need to have `Time`. The API isn't great but otherwise the `Time` is ... well in the past ;)

Reply:
Ok. Easy change. Just so I'm clear; which time should this be set to? It appears to be the start time of the request for websockets. Is that correct?

Reply:
yes, I think it is always the `start` time :D 

Unfortunately, I am unlikely to look into this more until next week ... just wanted to check something really fast, and ... this turned out to be a big problem :D